### PR TITLE
Keep track of failed task count

### PR DESCRIPTION
**Pull request by brandonhilkert**
In hopes to provide a more metrics around Sucker Punch jobs, I'm hoping
to expose the number of failed jobs in the pool.

Failed jobs would be visible in the chart on the left and potentially in the chart if I can find a way to not make it too busy:

screen shot 2015-10-18 at 9 11 46 pm

### CODE REVIEW

File: lib/concurrent/executor/ruby_thread_pool_executor.rb [hunk 1]
```diff
@@ -101,6 +106,11 @@ def worker_died(worker)
       synchronize { ns_worker_died worker }
     end
 
+    # @!visibility private
+    def worker_task_failed
+      @failed_task_count += 1
```
Comment (jdantonio) [inline lib/concurrent/executor/ruby_thread_pool_executor.rb]: This needs synchronized as well.

**Comment by jdantonio** [comment]
I really like the intent of this PR, but we need to look at the JRuby side. We have a strict rule that we don't ever build platform-specific features, only platform-specific optimizations. Since Java's [ThreadPoolExecutor](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html) doesn't expose this information we will need to track it in Ruby. I think that's workable.

`JavaExecutorService` defines an internal, private class called `Job` that calls the task block. If we added a `rescue` in there we could increment a counter. The (small) complexity is that we want the counter in `JavaThreadPoolExecutor` only, not `JavaExecutorService` (for compatibility with the Ruby implementation). So would could catch the error in `Job`, make a callback to the parent class, then re-raise the error. The default callback in `JavaExecutorService` would be a noop, which we could override in `JavaThreadPoolExecutor`. The latter would use an `AtomicFixnum` for the counter to make it thread safe.

Thoughts?

**Comment by brandonhilkert** [comment]
@jdantonio I'm not opposed to exploring, but based on your previous suggestion for `@ready_length`, I [implemented something quick](https://github.com/brandonhilkert/sucker_punch/blob/concurrent-web/lib/sucker_punch/job.rb#L37) and it's working well. I'm not opposed to doing the same for failed count in a `rescue` block. You tell if this has any value to the project because it would be just as easy for me to keep it in SP.

**Comment by jdantonio** [comment]
Cool. Since this feature would be a little messy on JRuby I'd prefer to hold off for now and leave it to the app to track. Thread pools are intended to be pretty low level and fast. Thanks for the input. I'm really excited to see how Sucker Punch evolves!

**Pull request closed without merging**